mempool/src/shared_mempool/network.rs

```diff
@@ -318,149 +335,190 @@ impl MempoolNetworkInterface {
         }
     }
 
-    pub fn execute_broadcast<V>(
+    /// Peers are prioritized when the local is a validator, or it's within the default failovers.
+    /// One is added for the primary peer
+    fn check_peer_prioritized(&self, peer: PeerNetworkId) -> Result<(), BroadcastError> {
+        if !self.role.is_validator() {
+            let priority = self
+                .prioritized_peers
+                .lock()
+                .iter()
+                .find_position(|peer_network_id| *peer_network_id == &peer)
+                .map_or(usize::MAX, |(pos, _)| pos);
+            if priority > self.mempool_config.default_failovers {
+                return Err(BroadcastError::PeerNotPrioritized(peer, priority));
+            }
+        }
+        Ok(())
+    }
+
+    /// Determines the broadcast batch.  There are three types of batches:
+    /// * Expired -> This timed out waiting for a response and needs to be resent
+    /// * Retry -> This received a response telling it to retry later
+    /// * New -> There are no Expired or Retry broadcasts currently waiting
+    fn determine_broadcast_batch<V>(
         &self,
         peer: PeerNetworkId,
         scheduled_backoff: bool,
         smp: &mut SharedMempool<V>,
-    ) where
+    ) -> Result<(BatchId, Vec<SignedTransaction>, Option<&str>), BroadcastError>
+    where
         V: TransactionValidation,
     {
-        // Start timer for tracking broadcast latency.
-        let start_time = Instant::now();
-
         let mut sync_states = self.sync_states.write_lock();
         let state = if let Some(state) = sync_states.get_mut(&peer) {
             state
         } else {
             // If we don't have any info about the node, we shouldn't broadcast to it
-            return;
+            return Err(BroadcastError::PeerNotFound(peer));
         };
 
-        // When not a validator, only broadcast to `default_failovers`
-        if !self.role.is_validator() {
-            let priority = self
-                .prioritized_peers
-                .lock()
-                .iter()
-                .find_position(|peer_network_id| *peer_network_id == &peer)
-                .map_or(usize::MAX, |(pos, _)| pos);
-            if priority > self.mempool_config.default_failovers {
-                return;
-            }
-        }
+        // If the peer isn't prioritized, lets not broadcast
+        self.check_peer_prioritized(peer)?;
```

Comment:
How will this affect the logs? i.e., in the old version of the code we're just returning. But now, we're returning an error (as a way to short circuit the call). Will the error be logged by the caller of this function? If so, is it likely to result in spam?

Reply:
They're logged by the caller of the function.  Yes, it could cause spam (but it's at trace level).

Reply:
Aah, sweet, thanks! I didn't follow the code path all the way back so wasn't sure how it was being logged 😄